### PR TITLE
WIP: networking.services: Add vpnc module

### PR DESCRIPTION
**Pull request by MasseR**
###### Motivation for this change

Related to the issue #77891, we have been using this vpnc module definition for a while now.

This MR creates a new services.networking.vpnc attributeset, for setting up a vpnc service and systemd definition (the latter was missing in the original module definition).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by MasseR** [comment]
Like mentioned above, I have been using this definition for a while now, but it's probably not super polished, so keep that in mind when reviewing

File: nixos/modules/services/networking/vpnc.nix [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        IPSec ID ${value.id}
+        IPSec secret ${value.secret}
+        Xauth username ${value.username}
+        Password helper ${pkgs.systemd}/bin/systemd-ask-password
+        Script ${pkgs.vpnc}/etc/vpnc/vpnc-script
+        '';
+    };
+  };
+  up = pkgs.writeScript "up" ''
+    #!${pkgs.bash}/bin/bash
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: stdenv.shell

File: nixos/modules/services/networking/vpnc.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    PATH=${makeBinPath [pkgs.nettools pkgs.iproute]}:$PATH
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: please set `path` on the unit.

File: nixos/modules/services/networking/vpnc.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    LOCAL_GATEWAY=$(ip r | grep "default via" | grep -oE "([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)" | head -n1)
+    ip route del default via $LOCAL_GATEWAY
+    ip route add default dev tun0 proto static scope link metric 50
+    if [ ! -d /var/run/vpnc ]; then
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: `/var/run` is deprecated. Also, use `RuntimeDirectory=vpnc` to create that directory

File: nixos/modules/services/networking/vpnc.nix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      mkdir -p /var/run/vpnc
+    fi
+    echo -ne $LOCAL_GATEWAY > /var/run/vpnc/LOCAL_GATEWAY
+  '';
+  down = pkgs.writeScript "down" ''
+    #!${pkgs.bash}/bin/bash
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: stdenv.shell, same as above.

File: nixos/modules/services/networking/vpnc.nix [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    PATH=${makeBinPath [pkgs.nettools pkgs.iproute]}:$PATH
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: set path in the unit, same as above

File: nixos/modules/services/networking/vpnc.nix [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    ip route del default dev tun0
+  '';
+  buildService = key: value: nameValuePair "vpnc-${key}" {
+    description = "VPNC for ${key}";
+    # No automatic startup, password
+
+    path = [ pkgs.iptables pkgs.nettools pkgs.iproute ];
+    serviceConfig = {
+      ExecStart = "${pkgs.vpnc}/bin/vpnc /etc/vpnc/${key}.conf";
+      ExecStartPost = "${up}";
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: you can pass `up` directly, without doing the string interpolation,

File: nixos/modules/services/networking/vpnc.nix [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      };
+      id = mkOption {
+        default = "linux";
+        type = types.separatedString "";
+        description = "Your group name";
+      };
+      secret = mkOption {
+        default = "";
+        type = types.separatedString "";
+        description = "Your shared secret";
```
Comment (flokli) [inline nixos/modules/services/networking/vpnc.nix]: This shouldn't be part of the nix store, but instead a path where things are read from at runtime.

Maybe it's better to do that for the whole config file - I assume ipsec supports a lot of more options as well…

**Review by flokli — CHANGES_REQUESTED** [review]
Also, see https://github.com/NixOS/nixpkgs/pull/74035 for the reasons this was dropped.

File: nixos/modules/services/networking/vpnc.nix (hunk 1 above)
Comment (Infinisil) [reply nixos/modules/services/networking/vpnc.nix]: Even better is `pkgs.runtimeShell`. `stdenv.shell` is not correct when cross-compiling

File: nixos/modules/services/networking/vpnc.nix (hunk 4 above)
Comment (Infinisil) [reply nixos/modules/services/networking/vpnc.nix]: `pkgs.runtimeShell` here too

**Comment by MasseR** [comment]
Thank you for the comments. I'll mark this as WIP until I get these implemented

File: nixos/modules/services/networking/vpnc.nix (hunk 7 above)
Comment (ckauhaus) [reply nixos/modules/services/networking/vpnc.nix]: I second that. Never store secrets in the Nix store (unless hashed). Better point to a file.

**Comment by flokli** [comment]
@MasseR do you plan to address these remarks in the near future? Otherwise, I'd just propose to close the PR for now, and reopen a new one once ready.

**Comment by MasseR** [comment]
Closing it for now

**Pull request closed without merging**